packages/uma-auth-client/src/hooks/useOAuth.tsx

```diff
@@ -0,0 +1,203 @@
+
+  // TODO: get requiredCommands, optionalCommands, budget, expirationPeriod from client app
+  const requiredCommands = ["pay_invoice"];
+  const optionalCommands = ["get_balance", "list_transactions"];
+  const budget = "10.USD%2Fmonthly";
+  const expirationPeriod = "year";
+
+  const authUrl = await authClient.generateAuthURL({
+    authorizeUrl: `${discoveryDocumentJson.authorization_endpoint}?required_commands=${requiredCommands.join(",")}&optional_commands=${optionalCommands.join(",")}&budget=${budget}&expiration_period=${expirationPeriod}`,
+    code_challenge_method: "S256",
+  });
+
+  return {
+    codeVerifier: authClient.code_verifier || "",
+    authUrl,
+  };
+};
+
+const oAuthTokenExchange = async (state: OAuthState, uma: string) => {
+  const {
+    identityNpub,
+    identityRelayUrl,
+    redirectUri,
+    codeVerifier,
+    token,
+    nwcConnectionUri,
+    nwcExpiresAt,
+  } = state;
+
+  // If we have a connection URI and a token that hasn't expired, we don't need to do anything
+  if (
+    nwcConnectionUri &&
+    token &&
+    (!token.expiresAt || Date.now() < token.expiresAt)
+  ) {
+    return state;
+  }
+
+  const authClient = new auth.OAuth2User({
+    client_id: `${identityNpub} ${identityRelayUrl}`,
+    callback: redirectUri!, // TODO: get this from client app
+    scopes: [],
+    user_agent: "uma-connect",
+    request_options: {
+      base_url: getNwcUmaDomain(uma),
+    },
+  });
+
+  let resultToken: UmaAuthToken;
+  if (token?.refreshToken && nwcExpiresAt && Date.now() < nwcExpiresAt) {
+    authClient.token = {
+      access_token: token.accessToken,
+      refresh_token: token.refreshToken,
+      expires_at: token.expiresAt,
+    };
+
+    resultToken = (await authClient.refreshAccessToken()).token as UmaAuthToken;
+  } else {
+    if (!codeVerifier) {
+      throw new Error(
+        "Code verifier not found. Make initial OAuth request first.",
+      );
+    }
+    authClient.code_verifier = codeVerifier;
+
+    // TODO: get code from client app in cases where they change URL params
+    const code = new URLSearchParams(window.location.search).get("code");
+    if (!code) {
+      throw new Error("Code not found in URL params");
+    }
+
+    resultToken = (await authClient.requestAccessToken(code))
+      .token as UmaAuthToken;
+  }
+  validateUmaAuthToken(resultToken);
+
+  return {
+    codeVerifier: "",
+    token: {
+      accessToken: resultToken.access_token,
+      refreshToken: resultToken.refresh_token,
+      expiresAt: resultToken.expires_at,
+    },
+    nwcConnectionUri: resultToken.nwc_connection_uri,
+    commands: resultToken.commands,
+    budget: resultToken.budget,
+    nwcExpiresAt: resultToken.nwc_expires_at,
+  };
+};
+
+const validateUmaAuthToken = (token: UmaAuthToken) => {
+  if (
+    !token.access_token ||
+    !token.refresh_token ||
+    !token.expires_at ||
+    !token.nwc_connection_uri ||
+    !token.commands ||
+    !token.budget ||
+    !token.nwc_expires_at
```

Comment:
I guess in theory these could be undefined, right?

Reply:
Oh just budget and nwc_expires_at? or are there other ones that could be undefined?

Reply:
Yeah I think just those two afaict.